### PR TITLE
Refactor missing namespace version

### PR DESCRIPTION
**Pull request by rly**
Based on [commentary](https://github.com/hdmf-dev/hdmf/commit/d432709374c7e8435f3a47fe29f906495d71ae97#r37363245) from @yarikoptic and @ajtritt, refactor, and:
1. expose the magic value that represents a missing version: None
2. write "None" as the version when writing the spec to disk
3. set the `SpecNamespace.version` to None if "None" is read as the version from disk
4. set the `SpecNamespace.version` to None if "unversioned" is read as the version from disk
5. make sure the "unversioned" and "None" version namespaces are not loaded when a versioned namespace is present
6. add a bunch of tests to cover these edge-cases and backward-compatibility support

### CODE REVIEW

**Review by ajtritt — DISMISSED** [review]

File: src/hdmf/spec/namespace.py [hunk 1]
```diff
@@ -83,13 +85,16 @@ def contact(self):
...
 
     @property
     def version(self):
-        """String, list, or tuple with the version or None """
-        return self.get('version', None)
+        """String, list, or tuple with the version or SpecNamespace.UNVERSIONED """
+        version = self.get('version', None)
+        if version is None:
+            self['version'] = SpecNamespace.UNVERSIONED
+        return self.get('version')
```
Comment (yarikoptic) [inline src/hdmf/spec/namespace.py]: `return self.get('version', None) or SpecNamespace.UNVERSIONED` probably do the same if version can't be numeric ;-) could also benefit of it is an empty list or an empty string for some reason

Comment (rly) [reply src/hdmf/spec/namespace.py]: The code as written also sets self['version']. How about changing line 95 to:
```python
if not version:
```

Comment (yarikoptic) [reply src/hdmf/spec/namespace.py]: I love functions with side effects, they are so much fun to debug ;-)
If anything, could also just not use version variable - used only once

Comment (rly) [reply src/hdmf/spec/namespace.py]: True, side effects are not great. I'll go with your suggestion.

File: src/hdmf/spec/namespace.py [hunk 2]
```diff
@@ -83,13 +85,16 @@ def contact(self):
...
 
     @property
     def version(self):
-        """String, list, or tuple with the version or None """
-        return self.get('version', None)
+        """
+        String, list, or tuple with the version or SpecNamespace.UNVERSIONED
+        if the version is missing or empty
+        """
+        return self.get('version', None) or SpecNamespace.UNVERSIONED
```
Comment (ajtritt) [inline src/hdmf/spec/namespace.py]: This would return a bool. Did you mean 

`return self.get('verzion', SpecNamespace.UNVERSIONED)`

Comment (ajtritt) [inline src/hdmf/spec/namespace.py:101]: This would return a bool. Did you mean this?

`return self.get('version', SpecNamespace.UNVERSIONED)`

File: src/hdmf/backends/hdf5/h5tools.py [hunk 3]
```diff
@@ -280,10 +280,7 @@ def write(self, **kwargs):
             for ns_name in ns_catalog.namespaces:
                 ns_builder = self.__convert_namespace(ns_catalog, ns_name)
                 namespace = ns_catalog.get_namespace(ns_name)
-                if namespace.version is None:
-                    group_name = '%s/unversioned' % ns_name
-                else:
-                    group_name = '%s/%s' % (ns_name, namespace.version)
+                group_name = '%s/%s' % (ns_name, namespace.version)
```
Comment (ajtritt) [inline src/hdmf/backends/hdf5/h5tools.py:294]: I changed the default to be `None`. This would make the group name "None". I don't have a problem with that, since I don't think it would impact reading the namespace

**Review by ajtritt — COMMENTED** [review]
@rly I made a couple of changes and comments. Please look them over... I suspect @yarikoptic would give a thumbs up.

File: src/hdmf/spec/namespace.py (hunk 2 above)
Comment (rly) [reply src/hdmf/spec/namespace.py:101]: Interestingly, this does not return a bool. 
```python                            
>>> 'a' or 'b'
'a'
>>> False or 'b'
'b'
>>> None or 'b'
'b'
>>> None or None
>>>                      
```
As written, if `version` is somehow '', [], or (), then `SpecNamespace.UNVERSIONED` will be returned. 

If `self.get('version', SpecNamespace.UNVERSIONED)` is used, then if `version` is somehow '', [], or (), then those would get returned.

File: src/hdmf/backends/hdf5/h5tools.py (hunk 3 above)
Comment (yarikoptic) [reply src/hdmf/backends/hdf5/h5tools.py:294]: cool.  When it is loaded from a file, would `"None"` string be casted to a proper `None` for the `.version`? Just checking, didn't spot anything obvious in the diff

Comment (rly) [reply src/hdmf/backends/hdf5/h5tools.py:294]: This is now resolved in `SpecNamespace.__init__`

Comment (rly) [reply src/hdmf/backends/hdf5/h5tools.py:294]: @yarikoptic Thanks for checking that!

**Review by ajtritt — APPROVED** [review]
